test/topology/util.py

```diff
@@ -74,13 +74,14 @@ async def check_token_ring_and_group0_consistency(manager: ManagerClient) -> Non
         assert token_ring_ids == group0_ids
 
 
-async def wait_for_token_ring_and_group0_consistency(manager: ManagerClient, deadline: float) -> None:
+async def wait_for_token_ring_and_group0_consistency(manager: ManagerClient, timeout: int = 30) -> None:
     """Weaker version of the above check; the token ring is not immediately updated after
     bootstrap/replace/decommission - the normal tokens of the new node propagate through gossip.
     Take this into account and wait for the equality condition to hold, with a timeout.
     """
     servers = await manager.running_servers()
     for srv in servers:
+        deadline = time.time() + timeout
```

Comment:
Apparently 30s appears to be too short for debug build, but @kbr-scylla said:
> In raft-topology mode they should be in sync immediately. If they aren't it means there is a bug in raft-topology

So maybe we need to address that.

Reply:
@kbr-scylla, the incident I reported about in the patch change log shiws that group0 and gossip aren't in immediate sync in this scenario.
The question is this an invariant we want yo maintain from now on.
Ifcthis is the case we shouldn't tolerate the inconsistency.

Reply:
Another question is whether the wait_for would eventually pass in the test run if it had a greater timeout.

I suspect it wouldn't, that there's either a bug in raft topology or we're making some wrong assumptions.

Note that `wait_for_token_ring_and_group0_consistency` assumes that group 0 configuration is immediately reflecting the latest operation, but allows the token ring to converge over time. This is because the test was written for gossip-topology mode, in which token ring was propagated over gossip but group 0 reconfigurations were done immediately during bootstrap/decommission/removenode/replace operations.

So the function fetches group 0 configuration **once** at the beginning, and fetches the token ring every `period`.

However, now we're running the tests in raft-topology mode. If I understand the coordinator code correctly, it's actually possible -- at least for the replace operation -- that we're removing a replaced node from group 0 configuration **after** we "finish" the operation (i.e. the replacing node finishes bootstrapping), so it is possible to fetch group 0 configuration **before** the replaced node was removed from it. So the situation reverses -- token ring is immediately consistent, but group 0 configuration may take some time to reflect the latest operation :)

If that is the case, we'd either need to fix the `wait_for_token_ring_and_group0_consistency` function to fetch group 0 periodically, or fix the topology coordinator code so it removes the replaced node from group 0 configuration before it considers the operation finished.

Anyway, @patjed41 is currently investigating the test_replace failures deeper, we'll see what his findings are.

TBH I don't think we need to increase the timeouts.

Reply:
@kbr-scylla what's the verdict?

Reply:
> @kbr-scylla what's the verdict?

There were two types of failures that I recall:
- one was basically a bug in Scylla causing group 0 and token ring to never converge. Already fixed by d1d1b6cf6e012143511a91e0170d7d1c6d89ffaa.
- another is that timeouts are too low because of driver sleeping: #15104

This PR wouldn't solve either. I sent a separate fix for #15104: #15105 that just increases the timeout from 30 to 60 seconds.

I think we can close this PR. As @tgrabiec observed, there's no point in picking timeouts sequentially for each server because the condition that we're checking is satisfied concurrently on all servers.

Reply:
okay